src/box/lua/load_cfg.lua

```diff
@@ -962,13 +964,40 @@ local function load_cfg(cfg)
             box[k] = v
         end
     end
+
     setmetatable(box, nil)
-    box.cfg = setmetatable(cfg,
+    -- There is only a single way for setting a value to a box option -
+    -- call 'box.cfg{ option_name=option_value }'. Direct modifications
+    -- of table box.cfg are not allowed. On attempt to make a direct
+    -- modifications (setting a new option or updating an existed one) we raise
+    -- an error. For implementation this we are using metamethods __newindex
+    -- and __index. Both __index and __newindex are relevant only when the
+    -- index does not exist in the table. The only way to catch all accesses to
+    -- a table is to keep it empty. So we use an empty proxy table and proxying
+    -- setting and reading to another table.
+    raw_cfg = table.copy(cfg)
+    local proxy_table = {}
+    box.cfg = setmetatable(proxy_table,
         {
-            __newindex = function(table, index) -- luacheck: no unused args
+            -- tarantoolctl (extra/dist/tarantoolctl.in) expects a metatable
+            -- with metamethods, not a string. Thus we couldn't use __metatable
+            -- instead.
+            __name = 'box_cfg',
+            __index = raw_cfg,
+            __newindex = function(self, key) -- luacheck: no unused args
+                if template_cfg[key] ~= nil then
+                    error('Use function box.cfg for modifications')
+                end
                 error('Attempt to modify a read-only table')
             end,
-            __call = locked(reload_cfg),
+            __call = function(self, ...)
+                -- assert(next(proxy_table) ~= nil,
+                --        "FATAL: length of proxy table is not zero")
+                return locked(reload_cfg)(raw_cfg, ...)
+            end,
+            __serialize = function()
+                return raw_cfg
+            end,
```

Comment:
Great we finally going to fix this issue! I have a couple of concerns though.

1. We have config options of type table. Examples are `wal_ext` or `log_level_modules`. Can we make table proxies recursively so that changing nested tables directly is also prohibited?

2. We probably better make same proxy for log.cfg in this patch or this patch series too. log.cfg have same issue.

Reply:
Looks like we can just apply same approach with empty proxy table recursively on box.cfg. This way we make box.cfg readonly on all levels. Then we can put this functionality info function and apply to log.cfg too.

Reply:
Discussed f2f and decided to create a new task for the above issues.